resolwe_bio/kb/management/commands/insert_mappings.py

```diff
@@ -34,26 +36,57 @@ def add_arguments(self, parser):
 
     def handle(self, *args, **options):
         """Command handle."""
-        count_inserted, count_unchanged = 0, 0
-
-        for tab_file_name, line_count, tab_file in decompress(options['file_name']):
-            logger.info(__("Importing mappings from \"{}\":", tab_file_name))
-
-            reader = csv.DictReader(tab_file, delimiter=str('\t'))
-            bar_format = '{desc}{percentage:3.0f}%|{bar}| {n_fmt}/{total_fmt} [{elapsed}<{remaining}]'
-
-            for row in tqdm(reader, total=line_count, bar_format=bar_format):
-                _, created = Mapping.objects.update_or_create(relation_type=row['relation_type'],
-                                                              source_db=row['source_db'],
-                                                              source_id=row['source_id'],
-                                                              target_db=row['target_db'],
-                                                              target_id=row['target_id'])
-                if created:
-                    count_inserted += 1
-                else:
-                    count_unchanged += 1
-
-        index_builder.push(index=MappingSearchIndex)
-
-        logger.info("Total mappings: %d. Inserted %d, unchanged %d." %  # pylint: disable=logging-not-lazy
-                    (count_inserted + count_unchanged, count_inserted, count_unchanged))
+        count_total = 0
+        to_index = []
+
+        relation_type_choices = list(zip(*Mapping.RELATION_TYPE_CHOICES))[0]
+
+        for tab_file_name, tab_file in decompress(options['file_name']):
+            logger.info(__("Importing mappings from \"{}\"...", tab_file_name))
+
+            mappings = []
+            unique_mappings = set()
+            for row in csv.DictReader(tab_file, delimiter=str('\t')):
+                if row['relation_type'] not in relation_type_choices:
+                    raise ValidationError(
+                        "Unknown relation type: {}".format(row['relation_type'])
+                    )
+
+                mapping = (
+                    row['relation_type'],
+                    row['source_db'],
+                    row['source_id'],
+                    row['target_db'],
+                    row['target_id'],
+                )
+
+                if mapping in unique_mappings:
+                    raise ValidationError("TODO")
+
+                mappings.append(mapping)
+                unique_mappings.add(mapping)
+
+            with connection.cursor() as cursor:
+                cursor.execute(
+                    """
+                    INSERT INTO {table_name} (
+                        relation_type, source_db, source_id, target_db, target_id
+                    )
+                    SELECT
+                        value->>0, value->>1, value->>2, value->>3, value->>4
+                    FROM json_array_elements(%s)
+                    ON CONFLICT DO NOTHING
```

Comment:
No updates for mappings?

Reply:
We have `unique_together` constraint that includes all fields: https://github.com/genialis/resolwe-bio/blob/master/resolwe_bio/kb/models.py#L97
So conflict means that exact same mapping is already present and we don't need to update anything.

Maybe I can rewrite this to `ON CONFLICT (relation_type, source_db, source_id, target_db, target_id) DO NOTHING` to make it clearer, but I have to test how this affects performance.

Reply:
I see, ok then just add a comment why this is so.